src/gerbil/runtime/c3.ss

```diff
@@ -1,84 +1,181 @@
 ;;;
+;;; We dub C4 our extension with the 4th constraints for single-inheritance.
+;;;
 ;;; See tests in ../test/c3-test.ss
 
 prelude: "../prelude/core"
 package: gerbil/runtime
 namespace: #f
 
-(export c3-linearize)
+(export c4-linearize c3-linearize)
 (import "util")
 
-;; C3 linearization algorithm: given a top object x from which to compute the precedence list,
+;; C4 linearization algorithm: given a top object x from which to compute the precedence list,
 ;; - rhead is [x] or [] depending on whether to include x as head of the result.
 ;; - supers is (get-supers x), the list of direct supers of x
 ;; - get-precedence-list gets the precedence list for a super s, including s itself in front
+;; - single-inheritance? is a predicate that tells if a class follows single inheritance, and
+;;   must have its precedence list be a tail of any subclass'es precedence list.
 ;; - eqpred is an equality predicate between list elements
 ;;   (passed around because somehow equal? doesn't work correctly on runtime type-descriptors).
 ;; - get-name gets the name of a object/class, for debugging only.
-;; : (List X) (List X) (X -> (NonEmptyList X)) ?(X X -> Bool) ?(X -> Y) -> (List X)
-(def (c3-linearize rhead supers get-precedence-list (eqpred eq?) (get-name identity))
-  (let (tails (map get-precedence-list supers)) ;; : (List (NonEmptyList X))
-    (append1! tails supers)
-    (c3-linearize-loop rhead tails eqpred get-name)))
+;; Returns the linearized precedence list, and the most specific struct superclass if any
+;; : (List X) (List X) \
+;;  get-precedence-list: (X -> (NonEmptyList X)) \
+;;  single-inheritance?: (X -> Bool) \
+;;  eqpred: ?(X X -> Bool) \
+;;  get-name: ?(X -> Y) \
+;; -> (List X) (OrFalse X)
+(def (c4-linearize rhead supers
+                   get-precedence-list: get-precedence-list
+                   single-inheritance?: single-inheritance?
+                   eqpred: (eqpred eq?)
+                   get-name: (get-name identity))
+  (def pls (map get-precedence-list supers)) ;; (List (List X)) ;; precedence lists to merge
+  (def sit []) ;; (List X) ;; single-inheritance tail
+
+  ;; Split every precedence list at the first struct, consider whatever
+  ;; follows as a tail of the precedence-list. Merge all the tails,
+  ;; where two tails are compatible if one is a tail of the other.
+  ;; Then in each remaining precedence list, (a) remove from the end the
+  ;; classes that are in the correct order in the tail, until you reach one
+  ;; that isn't in the tail, then check that no more class there is in the
+  ;; tail (or else there's an incompatibility).
+  ;; Use that as tail of the precedence list,
+  ;; and for the (reverse) head, proceed as usual with C3.
+
+  (def get-names (cut map get-name <>))
+  (def (err . a)
+    (apply error "Inconsistent precedence graph"
+           head: (get-names (reverse rhead))
+           precedence-lists: (map get-names pls)
+           single-inheritance-tail: (get-names sit) a))
+
```

Comment:
we should make this custom so that the compiler can raise a compile error that carries context; a plain error is not great for UX.